### DISCUSSION TITLE
Multiple ASP.NET Core Web API instances runs only once

### DISCUSSION THREAD

**Discussion by adryone** [dotnet/aspnetcore#59257] [Q&A] [answered]
I have an ASP.NET Core 8.0 Web API hosted on two IIS applications (app-1 and app-2) under the Default Web Site on a Windows 11 OS.

Both IIS applications point to the same physical path (`inetpub\wwwroot\myapp`) and each application has its own dedicated application pool (`app1 `and `app2`). The application pools have unique identities (`app1svc `and `app2svc`), both of which are members of the Administrators group.

In the Web API, I have an `AppEvents `class implementing `IHostedService`, with `StartAsync `and `StopAsync `methods to handle application start and stop events. In the `Program.cs`, I register it using `builder.Services.AddHostedService<AppEvents>()`.

When accessing `http://localhost/app-1`, the `StartAsync `method is triggered as expected. However, when accessing `http://localhost/app-2`, the `StartAsync `method does not execute. The WebApi application behind `http://localhost/app-1` or `http://localhost/app-2` starts only once, the first one that is requested.

I've tried changing the `AspNetHostingModel `from `InProcess `to `OutOfProcess`, but the behavior remains the same.

Is there a way to deploy multiple instances of the same web app, each running separately but pointing to the same physical directory, so that each instance correctly triggers its own `StartAsync`?``

**Comment by davidfowl** [comment]
@BrennanConroy ?

**Reply by adryone** [reply, marked as answer]
Hi David,

after several hours spent on the questiuon, I've found the "real" problerm. 
For monitoring I use NLog; the problem is that in the second "instance" of WebApi NLog does not write (there's a breaking change on NLog where the default of KeepFileOpen is changed from false to true).

Thanks
